src/libUtils/UpgradeManager.cpp

```diff
@@ -15,148 +15,307 @@
+        int line_no = 0;
+        string line;
+
+        /// Read major version
+        while (line_no != 2 && getline(versionFile, line))
+        {
+            ++line_no;
+        }
+
+        major = stoul(line);
+
+        /// Read minor version
+        while (line_no != 4 && getline(versionFile, line))
+        {
+            ++line_no;
+        }
+
+        minor = stoul(line);
+
+        /// Read fix version
+        while (line_no != 6 && getline(versionFile, line))
+        {
+            ++line_no;
+        }
+
+        fix = stoul(line);
+
+        /// Read expected DS epoch
+        while (line_no != 8 && getline(versionFile, line))
+        {
+            ++line_no;
+        }
 
+        upgradeDS = stoull(line);
+
+        /// Read Git commit ID
+        while (line_no != 12 && getline(versionFile, line))
+        {
+            ++line_no;
+        }
+
+        commit = stoul(line, nullptr, 16);
+
+        /// Read SHA-256 hash
+        while (line_no != 14 && getline(versionFile, sha))
+        {
+            ++line_no;
+        }
+    }
+
+    /// Verify SHA-256 checksum of .deb file
+    string downloadSha;
+    {
+        fstream debFile(debFileName, ios::in);
+
+        SHA2<HASH_TYPE::HASH_VARIANT_256> sha2;
+        vector<unsigned char> vec((istreambuf_iterator<char>(debFile)),
+                                  (istreambuf_iterator<char>()));
+        sha2.Update(vec, 0, vec.size());
+        vector<unsigned char> output = sha2.Finalize();
+        downloadSha = DataConversion::Uint8VecToHexStr(output);
+    }
+
+    if (sha != downloadSha)
+    {
+        LOG_GENERAL(WARNING, "SHA-256 checksum of .deb file does not match!");
+        return false;
+    }
+
+    m_latestSWInfo = make_shared<SWInfo>(major, minor, fix, upgradeDS, commit);
+    m_latestSHA = DataConversion::HexStrToUint8Vec(sha);
     return true;
 }
 
-bool UpgradeManager::ReplaceNode()
+bool UpgradeManager::ReplaceNode(Mediator& mediator)
 {
-    /// Store all the useful states into metadata, create a new node with loading the metadata, and kill current node
-    /// TBD
+    LOG_MARKER();
 
-    return true;
+    /// Store states
+    AccountStore::GetInstance().UpdateStateTrieAll();
+    AccountStore::GetInstance().MoveUpdatesToDisk();
+
+    /// Store DS block
+    vector<unsigned char> serializedDSBlock;
+    mediator.m_dsBlockChain.GetLastBlock().Serialize(serializedDSBlock, 0);
+    BlockStorage::GetBlockStorage().PutDSBlock(
+        mediator.m_dsBlockChain.GetLastBlock().GetHeader().GetBlockNum(),
+        serializedDSBlock);
+
+    /// Store final block
+    vector<unsigned char> serializedTxBlock;
+    mediator.m_txBlockChain.GetLastBlock().Serialize(serializedTxBlock, 0);
+    BlockStorage::GetBlockStorage().PutTxBlock(
+        mediator.m_txBlockChain.GetLastBlock().GetHeader().GetBlockNum(),
+        serializedTxBlock);
+
+    /// Deploy downloaded software
+    if (system("dpkg -i *.deb") < 0)
```

Comment:
Same `system()` issue here

Reply:
in addition, using *.deb is a dangerous practice too.

Reply:
Concretely, using *.deb presents at least two issues
1. you may install other non related packages
2. installation of the first packages may fail. However, if sub-sequence packages passed, I am worried that the return result might be 0 instead of 1 (to be verify). This may bypass some checks.

We should just avoid using * completely..